opt_adapt/opt.py

```diff
@@ -132,13 +148,162 @@ def _gradient_descent(it, forward_run, m, params, u, u_, dJ_, Rspace=False):
 
     # Take a step downhill
     u -= lr * dJ
-    yield {"lr": lr, "u+": u, "u-": u_, "dJ-": dJ_}
+    yield {"lr": lr, "u+": u}
+
+
+def _Adam(it, forward_run, m, params, u, a_, b_, Rspace=False, 
+        beta_1=0.9, beta_2=0.999, epsilon=1e-8,):
+    """
+    Take one iteration of the Adam algorithm.
+
+    :arg it: the current iteration number
+    :arg forward_run: a Python function that
+        implements the forward model and
+        computes the objective functional
+    :arg m: the current mesh
+    :kwarg params: :class:`OptAdaptParameters` instance
+        containing parameters for the optimisation and
+        adaptation routines
+    :arg u: the current control value
+    :arg a_: the previous first moment variable value
+    :arg b_: the previous second moment variable value
+    :kwarg Rspace: is the prognostic function
+        space of type 'Real'?
+    """
+
+    # Annotate the tape and compute the gradient
+    J, u = forward_run(m, u, **params.model_options)
+    dJ = fd_adj.compute_gradient(J, fd_adj.Control(u))
+    yield {"J": J, "u": u.copy(deepcopy=True), "dJ": dJ.copy(deepcopy=True)}
+
+    if a_ is None or b_ is None:
+        a_ = fd.Function(dJ.function_space())
+        b_ = fd.Function(dJ.function_space())
+
+    # Find the descent direction
+    if Rspace:
+        dJ_square = fd.Function(dJ).assign(pow(float(dJ), 2))
+        a = fd.Function(dJ).assign(beta_1 * a_ + (1 - beta_1) * dJ)
+        b = fd.Function(dJ).assign(beta_2 * b_ + (1 - beta_2) * dJ_square)
+        a_hat = fd.Function(dJ).assign(a / (1 - pow(beta_1, it)))
+        b_hat = fd.Function(dJ).assign(b / (1 - pow(beta_2, it)))
+        P = fd.Function(dJ).assign(-1 * a_hat / (pow(b_hat, 0.5) + epsilon))
+    else:
+        dJ_square = params.transfer_fn(dJ * dJ, dJ.function_space())
+        a = params.transfer_fn(beta_1 * a_ + (1 - beta_1) * dJ, dJ.function_space())
+        b = params.transfer_fn(
+            beta_2 * b_ + (1 - beta_2) * dJ_square, dJ.function_space()
+        )
+        a_hat = params.transfer_fn(a / (1 - pow(beta_1, it)), dJ.function_space())
+        b_hat = params.transfer_fn(b / (1 - pow(beta_2, it)), dJ.function_space())
+        P = params.transfer_fn(
+            -1 * a_hat / (pow(b_hat, 0.5) + epsilon), dJ.function_space()
+        )
```

Comment:
Note that applying `transfer_fn` to a UFL expression will only work in certain cases. However, it works for the default projection operator so we can probably leave it for now.